components/eamxx/src/physics/mam/eamxx_mam_optics_process_interface.cpp

```diff
@@ -1,110 +1,598 @@
+      m_params.isParameter("mam4_so4_physical_properties_file"),
+      "ERROR: mam4_so4_physical_properties_file is missing from mam_optics parameter list.");
+
+  EKAT_REQUIRE_MSG(
+      m_params.isParameter("mam4_pom_physical_properties_file"),
+      "ERROR: mam4_pom_physical_properties_file is missing from mam_optics parameter list.");
+
+  EKAT_REQUIRE_MSG(
+      m_params.isParameter("mam4_bc_physical_properties_file"),
+      "ERROR: mam4_bc_physical_properties_file is missing from mam_optics parameter list.");
 
-namespace scream
-{
+  EKAT_REQUIRE_MSG(
+      m_params.isParameter("mam4_mom_physical_properties_file"),
+      "ERROR: mam4_mom_physical_properties_file is missing from mam_optics parameter list.");
 
-MAMOptics::MAMOptics(
-    const ekat::Comm& comm,
-    const ekat::ParameterList& params)
-  : AtmosphereProcess(comm, params),
-    aero_config_() {
 }
 
 AtmosphereProcessType MAMOptics::type() const {
   return AtmosphereProcessType::Physics;
 }
 
-std::string MAMOptics::name() const {
-  return "mam4_optics";
-}
+std::string MAMOptics::name() const { return "mam4_optics"; }
 
-void MAMOptics::set_grids(const std::shared_ptr<const GridsManager> grids_manager) {
+void MAMOptics::set_grids(
+    const std::shared_ptr<const GridsManager> grids_manager) {
   using namespace ekat::units;
 
-  grid_ = grids_manager->get_grid("Physics");
-  const auto& grid_name = grid_->name();
+  grid_                 = grids_manager->get_grid("Physics");
+  const auto &grid_name = grid_->name();
+  auto q_unit           = kg / kg;  // mass mixing ratios [kg stuff / kg air]
+  q_unit.set_string("kg/kg");
+  auto n_unit = 1 / kg;  // number mixing ratios [# / kg air]
+  n_unit.set_string("#/kg");
+  const auto m2 = m * m;
+  const auto s2 = s * s;
 
-  ncol_ = grid_->get_num_local_dofs();      // number of columns on this rank
-  nlev_ = grid_->get_num_vertical_levels(); // number of levels per column
-  nswbands_ = 14;                           // number of shortwave bands
-  nlwbands_ = 16;                           // number of longwave bands
+  ncol_     = grid_->get_num_local_dofs();  // number of columns on this rank
+  nlev_     = grid_->get_num_vertical_levels();  // number of levels per column
+  nswbands_ = mam4::modal_aer_opt::nswbands;     // number of shortwave bands
+  nlwbands_ = mam4::modal_aer_opt::nlwbands;     // number of longwave bands
 
   // Define the different field layouts that will be used for this process
   using namespace ShortFieldTagsNames;
 
   // Define aerosol optics fields computed by this process.
   auto nondim = Units::nondimensional();
-  FieldLayout scalar3d_swband_layout { {COL, SWBND, LEV}, {ncol_, nswbands_, nlev_} };
-  FieldLayout scalar3d_lwband_layout { {COL, LWBND, LEV}, {ncol_, nlwbands_, nlev_} };
+  FieldLayout scalar3d_swbandp_layout{{COL, SWBND, ILEV},
+                                      {ncol_, nswbands_, nlev_ + 1}};
+
+  FieldLayout scalar3d_swbandp_layout2{{COL, SWBND, LEV},
```

Comment:
This naming convention is kind of confusing. Why is the layout for swband fields with vertical number LEV named swbandp_layout2 but for lwband it is lwband_layout?

Reply:
I just renamed  `scalar3d_swbandp_layout` to `scalar3d_swband_layout_int`. We need this layout, because mam4 uses an extra level for shortwaves aerosol fields, i.e.,ncol_, nswbands_, nlev_ + 1 .